csharp/src/Ice/InputStream.cs

```diff
@@ -1729,94 +1729,88 @@ internal MainEncapsBackup(Encaps encaps, int pos, Encoding encoding, int minTota
             }
         }
 
-        private sealed class Collection<T> : ICollection<T>
+        // Collection<T> holds the size of a Slice sequence and reads the sequence elements from the InputStream
+        // on-demand. It does not fully implement IEnumerable<T> and ICollection<T> (i.e. some methods throw
+        // NotSupportedException) because it's not resettable: you can't use it to unmarshal the same bytes
+        // multiple times.
+        private sealed class Collection<T> : ICollection<T>, IEnumerator<T>
         {
-            private readonly InputStream _ins;
-            private readonly InputStreamReader<T> _read;
-
             public int Count { get; }
 
+            public T Current
+            {
+                get
+                {
+                    if (_pos == 0 || _pos > Count)
+                    {
+                        throw new InvalidOperationException();
+                    }
+                    return _current;
+                }
+            }
+
+            object? IEnumerator.Current => Current;
             public bool IsReadOnly => true;
+            private T _current;
+            private readonly InputStream _inputStream;
+            private int _pos;
+            private readonly InputStreamReader<T> _reader;
 
-            public Collection(InputStream ins, InputStreamReader<T> read, int minSize)
+            // Disable this warning as the _current field is never read before it is initialized in MoveNext. Declaring
+            // this field as nullable is not an option for a genericT  that can be used with reference and value types.
+#pragma warning disable CS8618 // Non-nullable field is uninitialized. Consider declaring as nullable.
+            internal Collection(InputStream istr, InputStreamReader<T> reader, int minSize)
+#pragma warning restore CS8618
             {
-                _ins = ins;
-                _read = read;
-                Count = ins.ReadAndCheckSeqSize(minSize);
+                _inputStream = istr;
+                _reader = reader;
+                Count = istr.ReadAndCheckSeqSize(minSize);
             }
 
-            // TODO: Ideally this should use a InputStream view and cache the input stream start
-            // position, so that successive enumerators yield same valid results. In practice
-            // GetEnumerator should only be called once when the collection is unmarshaled.
-            public IEnumerator<T> GetEnumerator() => new Enumerator<T>(_ins, _read, Count);
+            public IEnumerator<T> GetEnumerator()
+            {
+                if (_pos != 0)
```

Comment:
This isn't very reliable, you can still create several enumerators, _pos will only be updated later when you use them, maybe we should use a separate flag to ensure this is only call once or throw.

Reply:
Fixed.